diff -urN xar/lib/darwinattr.c xar.patch/lib/darwinattr.c
--- xar/lib/darwinattr.c	2006-03-16 11:45:21.000000000 -0800
+++ xar.patch/lib/darwinattr.c	2006-03-16 11:48:27.000000000 -0800
@@ -569,10 +569,6 @@
 #if defined(__APPLE__)
 	if( len )
 		return 0;
-#if defined(HAVE_GETXATTR)
-	if( ea_archive(x, f, file, (void *)&context) == 0 )
-		return 0;
-#endif
 	if( nonea_archive(x, f, file, (void *)&context) == 0 )
 		return 0;
 	return underbar_archive(x, f, file, (void *)&context);
@@ -589,11 +585,6 @@
 #if defined(__APPLE__)
 	if( len )
 		return 0;
-#if defined(HAVE_GETXATTR)
-	if( ea_extract(x, f, file, (void *)&context) == 0 )
-		return 0;
-#endif
-
 	if( nonea_extract(x, f, file, (void *)&context) == 0 )
 		return 0;
 #endif /* __APPLE__ */
